### PR TITLE
Reuse previous results for better performance (replaces #49)

### PR DESCRIPTION
**Pull request by gerkey**
This is a replacement for #49 that fixes a couple of problems and adds tests to demonstrate that the in-process behavior is correct in the case of environment variable change: https://github.com/ros/rospack/blob/reuse_cache/test/test/utest.cpp#L130.

Requires ros/ros#98.

I think that this is a good change and am satisfied that we've done a good amount of testing on it. @dirk-thomas, what's your opinion? Other test cases that we should consider?

Btw, I'm targeting this change at indigo, but it could go to jade instead.

### CODE REVIEW

File: src/rospack.cpp [hunk 1]
```diff
@@ -2231,16 +2223,6 @@ Rospack::Rospack() :
...
-Rosstackage::~Rosstackage()
-{
-  for(std::tr1::unordered_map<std::string, Stackage*>::const_iterator it = stackages_.begin();
-      it != stackages_.end();
-      ++it)
-  {
-    delete it->second;
-  }
-}
-
```
Comment (dirk-thomas) [inline src/rospack.cpp:2243]: Please don't move the location of the destructor around. It is the same order as in the header and results in unnecessary changes.

File: include/rospack/rospack.h [hunk 2]
```diff
@@ -152,6 +152,7 @@ class ROSPACK_DECL Rosstackage
     std::tr1::unordered_map<std::string, Stackage*> stackages_;
     Stackage* findWithRecrawl(const std::string& name);
     void log(const std::string& level, const std::string& msg, bool append_errno);
+    void clearStackages();
```
Comment (dirk-thomas) [inline include/rospack/rospack.h:155]: This will break ABI.

Comment (gerkey) [inline include/rospack/rospack.h:155]: That also occurred to me, to, but I didn't think that we were ensuring ABI compatibility (instead counting on the rebuild-everything-and-version-lock-the-debs approach on the farm).

Comment (dirk-thomas) [inline include/rospack/rospack.h:155]: We don't enforce it but try to avoid it when possible since many more people have from-source builds.

Comment (rhaschke) [inline include/rospack/rospack.h:155]: I also was afraid of breaking ABI. However, [here](https://techbase.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B#The_Do.27s_and_Don.27ts) it was stated that

  _adding new non-virtual functions_

is no problem for ABI compatibility. This finding was supported by [several other sources](https://www.google.de/search?q=abi+compatibility+adding+method&oq=abi+&aqs=chrome.0.69i59j0l2j69i57j0l2.1878j0j7&client=ubuntu&sourceid=chrome&es_sm=122&ie=UTF-8) too.

Comment (dirk-thomas) [inline include/rospack/rospack.h:155]: Good point. I didn't think about non-virtual.

**Comment by gerkey** [comment]
Closing in favor of #49
